File: clients/typescript/src/validation.ts
// =============================================================================
// HYDRA-UMC-SDK - TypeScript reference client: real JSON Schema runtime validation
// Copyright (C) 2026 JuanenRac (Electro Hobby 3D) <electrohobby3d@gmail.com>
// GPL-3.0-or-later - see LICENSE
// =============================================================================

import * as fs from "node:fs";
import * as path from "node:path";
import Ajv2020, { type ErrorObject, type ValidateFunction } from "ajv/dist/2020";
import addFormats from "ajv-formats";

import type { ContractName } from "./types";

/**
 * Raised by `validate()` on the first schema violation Ajv reports, or on
 * an unknown contract name. A `ValueError`-shaped subclass, matching the
 * Python reference client's own `ContractValidationError`
 * (`clients/python/src/hydra_umc_sdk/validation.py`).
 */
export class ContractValidationError extends Error {
  readonly contract: string;
  readonly errors: ErrorObject[] | null | undefined;

  constructor(contract: string, message: string, errors?: ErrorObject[] | null) {
    super(`${contract}: ${message}`);
    this.name = "ContractValidationError";
    this.contract = contract;
    this.errors = errors;
  }
}

/**
 * Vendored schema file for each of the 7 real payload contracts this
 * client validates - the same 7 names documented in
 * docs/PYTHON_CLIENT.md's REQUIRED table for the Python reference client.
 */
const CONTRACT_FILES: Record<ContractName, string> = {
  DeviceDescriptor: "device-descriptor.schema.json",
  HealthReport: "health-report.schema.json",
  SafetyState: "safety-state.schema.json",
  UpdateManifest: "update-manifest.schema.json",
  EventEnvelope: "event-envelope.schema.json",
  ServerDiscovery: "server-discovery.schema.json",
  ProjectManifest: "project-manifest.schema.json",
};

/**
 * This package's own vendored schema copies (schemas/*.json, shipped
 * alongside dist/ via package.json's "files"). __dirname resolves relative
 * to the *compiled* file, so this stays correct whether run from src/
 * (ts-node/native TS) or dist/ (tsc output) - both sit directly under
 * clients/typescript/, one level above schemas/.
 */
const SCHEMA_DIR = path.join(__dirname, "..", "schemas");

const ajv = new Ajv2020({ allErrors: true, strict: true });
addFormats(ajv);

const compiled = new Map<ContractName, ValidateFunction>();

function compiledValidator(contract: ContractName): ValidateFunction {
  const cached = compiled.get(contract);
  if (cached) {
    return cached;
  }
  const file = CONTRACT_FILES[contract];
  if (!file) {
    throw new ContractValidationError(contract, "unknown contract");
  }
  const schemaPath = path.join(SCHEMA_DIR, file);
  let schema: unknown;
  try {
    schema = JSON.parse(fs.readFileSync(schemaPath, "utf-8"));
  } catch (error) {
    throw new ContractValidationError(contract, `cannot load schema ${schemaPath}: ${(error as Error).message}`);
  }
  const validator = ajv.compile(schema as Record<string, unknown>);
  compiled.set(contract, validator);
  return validator;
}

/**
 * Validate `payload` against the real, published v1 JSON Schema for
 * `contract`. Throws `ContractValidationError` on the first violation Ajv
 * reports (all errors are still collected on `.errors`); returns nothing
 * on a valid payload. This enforces the full schema (via Ajv, a real
 * draft 2020-12 JSON Schema validator with `ajv-formats` enabled for
 * `format: "date-time"`), not a hand-reimplemented subset.
 */
export function validate(contract: ContractName, payload: unknown): void {
  const validator = compiledValidator(contract);
  const ok = validator(payload);
  if (!ok) {
    const message = ajv.errorsText(validator.errors, { separator: "; " });
    throw new ContractValidationError(contract, message, validator.errors);
  }
}

/** `validate()` as a boolean predicate instead of a throw/no-throw call. */
export function isValid(contract: ContractName, payload: unknown): boolean {
  try {
    validate(contract, payload);
    return true;
  } catch {
    return false;
  }
}

/** The contract names `validate()`/`isValid()` accept, sorted. */
export function knownContracts(): ContractName[] {
  return (Object.keys(CONTRACT_FILES) as ContractName[]).sort();
}

File: clients/rust/tests/validation.rs
// =============================================================================
// HYDRA-UMC-SDK - Rust reference client tests: real schema validation
// Copyright (C) 2026 JuanenRac (Electro Hobby 3D) <electrohobby3d@gmail.com>
// GPL-3.0-or-later - see LICENSE
// =============================================================================

use std::fs;
use std::path::{Path, PathBuf};

use serde_json::{Value, json};

use hydra_umc_sdk::{known_contracts, validate, validate_json, verify_contract_manifest};

/// Locates `conformance/fixtures/v1` relative to this monorepo checkout -
/// the same fixtures `clients/python/tests`, `clients/go/validation_test.go`
/// and `clients/typescript/tests/validation.test.mts` all validate against.
fn fixtures_dir() -> Option<PathBuf> {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join("conformance")
        .join("fixtures")
        .join("v1");
    dir.is_dir().then_some(dir)
}

fn read_fixture(dir: &Path, name: &str) -> String {
    fs::read_to_string(dir.join(name))
        .unwrap_or_else(|error| panic!("cannot read fixture {name}: {error}"))
}

/// The same fixture-naming convention `tools/verify_contract_matrix.py`
/// enforces: `<kebab-case-contract>.valid.json` / `.invalid.json`.
const CONTRACT_FIXTURE_STEM: &[(&str, &str)] = &[
    ("DeviceDescriptor", "device-descriptor"),
    ("EventEnvelope", "event-envelope"),
    ("HealthReport", "health-report"),
    ("ProjectManifest", "project-manifest"),
    ("SafetyState", "safety-state"),
    ("ServerDiscovery", "server-discovery"),
    ("UpdateManifest", "update-manifest"),
];

#[test]
fn accepts_every_valid_fixture() {
    let Some(dir) = fixtures_dir() else {
        eprintln!(
            "conformance fixtures not found; skipping (expected inside a HYDRA-UMC-SDK checkout)"
        );
        return;
    };
    for (contract, stem) in CONTRACT_FIXTURE_STEM {
        let data = read_fixture(&dir, &format!("{stem}.valid.json"));
        assert!(
            validate_json(contract, &data).is_ok(),
            "expected {contract} valid fixture to pass"
        );
    }
}

#[test]
fn rejects_every_invalid_fixture() {
    let Some(dir) = fixtures_dir() else {
        eprintln!(
            "conformance fixtures not found; skipping (expected inside a HYDRA-UMC-SDK checkout)"
        );
        return;
    };
    for (contract, stem) in CONTRACT_FIXTURE_STEM {
        let data = read_fixture(&dir, &format!("{stem}.invalid.json"));
        assert!(
            validate_json(contract, &data).is_err(),
            "expected {contract} invalid fixture to be rejected"
        );
    }
}

#[test]
fn unknown_contract_is_rejected() {
    let error = validate("NotARealContract", &json!({})).unwrap_err();
    assert_eq!(error.contract, "NotARealContract");
}

#[test]
fn known_contracts_lists_all_seven() {
    let names = known_contracts();
    assert_eq!(names.len(), 7, "expected 7 known contracts, got {names:?}");
}

#[test]
fn rejects_boolean_for_integer_sequence() {
    // bool must never satisfy an "integer" schema - mirrors
    // clients/python/tests/test_validation.py's
    // test_rejects_boolean_event_sequence, clients/go's
    // TestValidateRejectsBooleanForIntegerSequence, and clients/typescript's
    // equivalent test.
    let payload: Value = json!({
        "schema_version": "1.0",
        "event_id": "e",
        "type": "x",
        "source": "os",
        "timestamp_utc": "2026-08-26T12:00:00Z",
        "sequence": true,
    });
    assert!(validate("EventEnvelope", &payload).is_err());
}

#[test]
fn rejects_invalid_date_time() {
    let payload: Value = json!({
        "schema_version": "1.0",
        "state": "READY",
        "timestamp_utc": "2026-02-30T12:00:00Z", // no such calendar day
        "checks": {},
    });
    assert!(validate("HealthReport", &payload).is_err());
}

#[test]
fn accepts_a_minimal_health_report() {
    let payload: Value = json!({
        "schema_version": "1.0",
        "state": "READY",
        "timestamp_utc": "2026-01-01T00:00:00Z",
        "checks": {"storage": {"state": "PASS"}},
    });
    assert!(validate("HealthReport", &payload).is_ok());
}

#[test]
fn rejects_unknown_property_on_project_manifest() {
    // project-manifest.schema.json is the one contract with
    // additionalProperties: false - an unknown field is a real violation,
    // not an allowed extension.
    let Some(dir) = fixtures_dir() else {
        eprintln!(
            "conformance fixtures not found; skipping (expected inside a HYDRA-UMC-SDK checkout)"
        );
        return;
    };
    let data = read_fixture(&dir, "project-manifest.valid.json");
    let mut payload: Value = serde_json::from_str(&data).expect("fixture must be valid JSON");
    payload
        .as_object_mut()
        .expect("fixture must be a JSON object")
        .insert("unexpected_field".to_string(), json!("nope"));
    assert!(validate("ProjectManifest", &payload).is_err());
}

#[test]
fn verify_contract_manifest_passes() {
    verify_contract_manifest()
        .expect("every embedded schema must match manifest.json's own sha256 digest");
}
